pkg/kds/store/sync.go

```diff
@@ -0,0 +1,139 @@
+package store
+
+import (
+	"context"
+
+	"github.com/Kong/kuma/pkg/core/resources/model"
+	"github.com/Kong/kuma/pkg/core/resources/registry"
+	"github.com/Kong/kuma/pkg/core/resources/store"
+)
+
+// SyncResourceStore extends ResourceStore with Sync method
+type SyncResourceStore interface {
+	store.ResourceStore
+	// Sync method takes 'upstream' as a basis and synchronize underlying store.
+	// It deletes all resources that absent in 'upstream', creates new resources that
+	// are not represented in store yet and updates the rest.
+	Sync(upstream model.ResourceList, fs ...SyncOptionFunc) error
+}
+
+type SyncOption struct {
+	Predicate func(r model.Resource) bool
+}
+
+type SyncOptionFunc func(*SyncOption)
+
+func NewSyncOptions(fs ...SyncOptionFunc) *SyncOption {
+	opts := &SyncOption{}
+	for _, f := range fs {
+		f(opts)
+	}
+	return opts
+}
+
+func PrefilterBy(predicate func(r model.Resource) bool) SyncOptionFunc {
+	return func(opts *SyncOption) {
+		opts.Predicate = predicate
+	}
+}
+
+type syncResourceStore struct {
+	store.ResourceStore
+}
+
+func NewSyncResourceStore(resourceStore store.ResourceStore) SyncResourceStore {
+	return &syncResourceStore{
+		ResourceStore: resourceStore,
+	}
+}
+
+func (s *syncResourceStore) Sync(upstream model.ResourceList, fs ...SyncOptionFunc) error {
+	opts := NewSyncOptions(fs...)
+	ctx := context.Background()
+
+	downstream, err := registry.Global().NewList(upstream.GetItemType())
+	if err != nil {
+		return err
+	}
+	if err := s.ResourceStore.List(ctx, downstream); err != nil {
+		return err
+	}
+
+	if opts.Predicate != nil {
+		filtered, err := registry.Global().NewList(upstream.GetItemType())
+		if err != nil {
+			return err
+		}
+		for _, r := range downstream.GetItems() {
+			if opts.Predicate(r) {
+				if err := filtered.AddItem(r); err != nil {
+					return err
+				}
+			}
+		}
+		downstream = filtered
+	}
+
+	// 1. delete resources from store which are not represented in 'upstream'
+	onDelete := []model.Resource{}
+	for _, r := range downstream.GetItems() {
+		if !set(upstream.GetItems()).contains(model.MetaToResourceKey(r.GetMeta())) {
+			onDelete = append(onDelete, r)
+		}
+	}
+
+	// 2. create resources which are not represented in 'downstream' and update the rest of them
+	onCreate := []model.Resource{}
+	onUpdate := []model.Resource{}
+	for _, r := range upstream.GetItems() {
+		if existing := set(downstream.GetItems()).get(model.MetaToResourceKey(r.GetMeta())); existing != nil {
+			// we have to use meta of the current store during update
+			r.SetMeta(existing.GetMeta())
+			onUpdate = append(onUpdate, r)
```

Comment:
Correct me if I'm wrong, but we are updating the resource regardless if anything changed. We should check if the spec is different.

Reply:
Yes, you are right, I'll add